roachtest: consider specific cloud config for reuse

nameisbhaskar opened the pull request:
We compare the entire spec for cluster reuse including the cloud specific config. This is wrong because if the cloud specific cluster spec differs, we end up creating a new cluster e.g. if we are running the test for GCE and the cluster spec differs for AWS, we can reuse the cluster for GCE.

Fixes: #125476
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125477)
<!-- Reviewable:end -->

vidit-bhat commented on pkg/cmd/roachtest/spec/cluster_spec.go:
```diff
@@ -153,9 +153,32 @@ func MakeClusterSpec(nodeCount int, opts ...Option) ClusterSpec {
 // ClustersCompatible returns true if the clusters are compatible, i.e. the test
 // asking for s2 can reuse s1.
-func ClustersCompatible(s1, s2 ClusterSpec) bool {
+func ClustersCompatible(s1, s2 ClusterSpec, cloud string) bool {
 	s1.Lifetime = 0
 	s2.Lifetime = 0
+	// only consider the specification of the cloud that we are running in
+	if cloud != GCE {
+		s1.GCE.VolumeType = ""
+		s1.GCE.Zones = ""
+		s1.GCE.MachineType = ""
+		s1.GCE.MinCPUPlatform = ""
+		s2.GCE.VolumeType = ""
+		s2.GCE.Zones = ""
+		s2.GCE.MachineType = ""
+		s2.GCE.MinCPUPlatform = ""
+	}
+	if cloud != AWS {
+		s1.AWS.Zones = ""
+		s1.AWS.VolumeThroughput = 0
+		s1.AWS.MachineType = ""
+		s2.AWS.Zones = ""
+		s2.AWS.VolumeThroughput = 0
+		s2.AWS.MachineType = ""
+	}
+	if cloud != Azure {
+		s1.Azure.Zones = ""
+		s2.Azure.Zones = ""
+	}
 	return s1 == s2
 }
```
How about
```
func clearAWS(c *ClusterSpec) {
	c.AWS.Zones = ""
	c.AWS.VolumeThroughput = 0
	c.AWS.MachineType = ""
}

func clearGCE(c *ClusterSpec) {
	c.GCE.VolumeType = ""
	c.GCE.Zones = ""
	c.GCE.MachineType = ""
	c.GCE.MinCPUPlatform = ""
}

func clearAzure(c *ClusterSpec) {
	c.Azure.Zones = ""
}

func clearClusterSpecFields(c *ClusterSpec, cloud string) {
	c.Lifetime = 0
	switch cloud {
	// only consider the specification of the cloud that we are running in
	case GCE:
		clearAWS(c)
		clearAzure(c)
	case AWS:
		clearGCE(c)
		clearAzure(c)
	case Azure:
		clearGCE(c)
		clearAWS(c)
	}
}

// ClustersCompatible returns true if the clusters are compatible, i.e. the test
// asking for s2 can reuse s1.
func ClustersCompatible(s1, s2 ClusterSpec, cloud string) bool {
	clearClusterSpecFields(&s1, cloud)
	clearClusterSpecFields(&s2, cloud)
	return s1 == s2
}
```
We could generalise the clear functions using reflection [1]

[1] https://stackoverflow.com/a/29169727

nameisbhaskar replied:
looks neater. will change this - Thanks!

herkolategan reviewed (commented):
In the commit msg / title`roachtest - consider specific cloud config for reuse`, using a dash is a bit strange, as the convention on PRs are `<packages>: <desc>` usually.

I'd probably want this PR to somehow make it clear that anyone adding something to the cluster spec struct needs to consider the reuse/ClusterCompatible aspect of it. Someone adding a new field at the moment won't even be aware they are messing up this logic.

srosenberg commented:
> We compare the entire spec for cluster reuse including the cloud specific config. This is wrong because if the cloud specific cluster spec differs, we end up creating a new cluster e.g. if we are running the test for GCE and the cluster spec differs for AWS, we can reuse the cluster for GCE.
> 

I am not following the last sentence. Do you mean that the comparison should extend to considering specs. across different cloud providers?

herkolategan reviewed (commented):
I wrote a small suggestion program as to how I think this can be done more cleanly with reflection, so that maintainers of the `ClusterSpec` struct do not have to update `ClustersCompatible` if the struct is amended. Although I do think a comment on the `ClusterSpec` struct somewhere that indicates the struct is compared for re-use might be a good idea as well.

```go
package main

import (
	"fmt"
	"reflect"
	"testing"
)

type ClusterSpec struct {
	GCE struct {
		MachineType string
		VolumeType  string
	} `cloud:"GCE"`
	AWS struct {
		MachineType string
		Zones       string
	} `cloud:"AWS"`
}

func TestClearClusterSpec(t *testing.T) {
	cs := ClusterSpec{
		GCE: struct {
			MachineType string
			VolumeType  string
		}{
			"gce_test", "gce_test",
		},
		AWS: struct {
			MachineType string
			Zones       string
		}{
			"aws_test", "aws_test",
		},
	}

	fmt.Printf("BEFORE: %+v\n", cs)

	structType := reflect.TypeOf(cs)
	for i := 0; i < structType.NumField(); i++ {
		// Get field
		field := structType.Field(i)
		// Check if the "cloud" tag exists
		if tag, ok := field.Tag.Lookup("cloud"); ok {
			// Zero out struct if not GCE (for case cloud == GCE)
			if tag != "GCE" {
				fieldValue := reflect.ValueOf(&cs).Elem().FieldByName(field.Name)
				fieldValue.Set(reflect.Zero(fieldValue.Type()))
			}
		}
	}
	fmt.Printf("AFTER: %+v\n", cs)
}
```

herkolategan commented on pkg/cmd/roachtest/spec/cluster_spec.go:
```diff
@@ -153,12 +153,46 @@ func MakeClusterSpec(nodeCount int, opts ...Option) ClusterSpec {
+	c.Lifetime = 0
+	switch cloud {
+	// only consider the specification of the cloud that we are running in
+	case GCE:
```
See my suggestion around how we can avoid clearing "other" clouds using reflection.

herkolategan commented:
One thing that also comes to mind with this PR is why we have `ClusterSpec`(s) that have the same definition for GCE, but different params for say AWS. Something about that doesn't feel right, and maybe more like those should be aligned - although I'm not sure how we would enforce something like that, and maybe there is a case for it where it makes sense?

nameisbhaskar committed:
roachtest: consider specific cloud config for reuse

We compare the entire spec for cluster reuse including the cloud specific config. This is wrong because if the cloud specific cluster spec defers and we are running the test in a different cloud, we end up creating a new cluster
 e.g. if we are running the test on GCE and the cluster spec differs for AWS, we can reuse the cluster for GCE.

Fixes: #125476
Epic: None

nameisbhaskar commented:
Thanks @herkolategan for the code! I have modified the code and added an UT.

nameisbhaskar commented:
> > We compare the entire spec for cluster reuse including the cloud specific config. This is wrong because if the cloud specific cluster spec differs, we end up creating a new cluster e.g. if we are running the test for GCE and the cluster spec differs for AWS, we can reuse the cluster for GCE.
> 
> I am not following the last sentence. Do you mean that the comparison should extend to considering specs. across different cloud providers?

Sorry for the confusion. It is the other way round. If a specification differs for GCE, cluster reusability check should pass for AWS.

herkolategan commented on pkg/cmd/roachtest/spec/cluster_spec.go:
```diff
@@ -153,12 +154,30 @@ func MakeClusterSpec(nodeCount int, opts ...Option) ClusterSpec {
+		field := structType.Field(i)
+		// Check if the "cloud" tag exists
+		if tag, ok := field.Tag.Lookup("cloud"); ok {
+			// Zero out struct if not GCE (for case cloud == GCE)
```
This comment may not be applicable anymore, might be good to update it to the say the `cloud` parameter. `zero out struct if not target cloud`

herkolategan commented on pkg/cmd/roachtest/spec/cluster_spec.go:
```diff
@@ -153,12 +154,30 @@ func MakeClusterSpec(nodeCount int, opts ...Option) ClusterSpec {
 	return s1 == s2
 }
 
+func clearClusterSpecFields(cs *ClusterSpec, cloud string) {
```
May be good to add a comment to this function explaining that it will clear everything except the passed cloud. The name of the function makes it sound a bit like it will clear the passed cloud, so a comment will make it clearer what's happening.

herkolategan commented on pkg/cmd/roachtest/spec/cluster_spec_test.go:
```diff
@@ -0,0 +1,44 @@
+		s1 := ClusterSpec{NodeCount: 5}
+		s2 := ClusterSpec{NodeCount: 5}
+		s1.GCE.VolumeType = "mock_volume1"
+		s1.GCE.VolumeType = "mock_volume2"
```
This test does not seem like it does what the description says, should it be `AWS` here?

herkolategan reviewed (changes requested):
Looks good, requesting a change for the one test that seems like it may need to be updated to prove correctness.

herkolategan commented on pkg/cmd/roachtest/spec/cluster_spec.go:
```diff
@@ -153,12 +154,33 @@ func MakeClusterSpec(nodeCount int, opts ...Option) ClusterSpec {
 }
 
+// clearClusterSpecFields clears the cloud specific specification from the cluster spec
+// if the cloud specification dos not match the target cloud. This is done to ensure that
```
*does

nameisbhaskar commented:
Thanks @herkolategan for the review and suggestions!

nameisbhaskar commented:
bors r=@herkolategan

srosenberg commented on pkg/cmd/roachtest/spec/cluster_spec.go:
```diff
@@ -153,12 +154,33 @@ func MakeClusterSpec(nodeCount int, opts ...Option) ClusterSpec {
-	s2.Lifetime = 0
+func ClustersCompatible(s1, s2 ClusterSpec, cloud string) bool {
+	// only consider the specification of the cloud that we are running in
+	clearClusterSpecFields(&s1, cloud)
```
Pardon my ignorance... why did we have to resort to a rather hacky way of implementing equality _modulo_? We're already using `cmp.Equal` in places, e.g., [1]. It's a much cleaner way, imo; specifically, it's side-effect free.

[1] https://github.com/cockroachdb/cockroach/blob/3bd657275574961162a702c9407f45f885fdd0e0/pkg/sql/sem/tree/backup.go#L390
